minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,796 @@
+    ) -> None:
+        if check_edge_valid:
+            if not isinstance(x, ZNode) or not isinstance(y, ZNode):
+                raise TypeError(f"Expected x, y to be ZNode, got {type(x), type(y)}")
+
+            if x.shape != y.shape:
+                raise ValueError(f"Expected x, y to have the same shape, got {x.shape, y.shape}")
+
+            for kind in EdgeKind:
+                if x.is_neighbor(y, nbr_kind=kind):
+                    self._edge_kind = kind
+                    break
+            else:
+                raise ValueError(f"Expected x, y to be neighbors, got {x, y}")
+
+        self._edge = self._set_edge(x, y)
+
+    @property
+    def edge_kind(self) -> EdgeKind:
+        return self._edge_kind
+
+
+class ZNode:
+    """Initializes 'ZNode' with coord and optional shape.
+
+    Args:
+        coord (CartesianCoord | ZephyrCoord | tuple[int]): Coordinate in (quotient) Zephyr or (quotient) Cartesian
+        shape (ZShape | tuple[int | None] | None, optional): Shape of the Zephyr graph containing this ZNode.
+            If a ZShape is passed, it should be a namedtuple with fields `m` (grid size of the Zephyr graph) and `t` (tile size of the Zephyr graph).
+        Defaults to None.
+        convert_to_z (bool | None, optional): Whether to express the coordinates in ZephyrCoordinates.
+            Defaults to None.
+
+    Note: 
+        If the `k` field of the given `coord` (whether a `CartesianCoord` or `ZephyrCoord`) is not `None`,
+        then `shape` must be provided and its `t` field (the tile size of the Zephyr graph) must not be `None`.
+        Otherwise, a `ValueError` will be raised.
+
+    Example:
+    >>> from zephyr_utils.node_edge import ZNode, ZShape
+    >>> zn1 = ZNode((5, 2), ZShape(m=5))
+    >>> zn1.neighbors()
+    [ZNode(CartesianCoord(x=4, y=1, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=4, y=3, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=6, y=1, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=6, y=3, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=1, y=2, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=9, y=2, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=3, y=2, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=7, y=2, k=None), shape=ZShape(m=5, t=None))]
+    >>> from zephyr_utils.node_edge import ZNode, ZShape
+    >>> zn1 = ZNode((5, 2), ZShape(m=5))
+    >>> zn1.neighbors(nbr_kind=EdgeKind.ODD)
+    [ZNode(CartesianCoord(x=3, y=2, k=None), shape=ZShape(m=5, t=None)),
+        ZNode(CartesianCoord(x=7, y=2, k=None), shape=ZShape(m=5, t=None))]
+    """
+
+    def __init__(
+        self,
+        coord: CartesianCoord | ZephyrCoord | tuple[int],
+        shape: ZShape | tuple[int | None] | None = None,
+        convert_to_z: bool | None = None,
+    ) -> None:
+        if shape:
+            self.shape = shape
```

Comment:
Since `ZNode.shape` is expected to always return a `ZShape`.
```suggestion
            self.shape = ZShape(*shape)
```